src/link.ts

```diff
@@ -0,0 +1,234 @@
+    right = ApolloLink.toLink(right);
+
+    if (ApolloLink.isTerminating(left) && ApolloLink.isTerminating(right)) {
+      return new TerminatedSplit(test, left, right);
+    }
+
+    return new SplitLink(test, left, right);
+  }
+
+  public static toLink(link: ApolloLink | RequestHandler): ApolloLink {
+    if (typeof link === 'function') {
+      return link.length <= 1 ? new TerminatedLink(link) : new FunctionLink(link);
+    } else {
+      return link as ApolloLink;
+    }
+  }
+
+  public static isTerminating(link: ApolloLink): boolean {
+    return link.request.length <= 1;
+  }
+
+  public split(
+    test: (op: Operation) => boolean,
+    left: ApolloLink | RequestHandler,
+    right: ApolloLink | RequestHandler = ApolloLink.passthrough(),
+  ): ApolloLink {
+    return this.concat(<ApolloLink>ApolloLink.split(test, left, right));
+  }
+
+  // join two Links together
+  public concat(link: ApolloLink | RequestHandler): ApolloLink {
+    if (this.request.length <= 1) {
+      const warning = Object.assign(
+        new Error(`You are calling concat a terminating link, which will have no effect`),
+        { link : this },
+      );
+      console.warn(warning);
+      return this;
+    }
+    link = ApolloLink.toLink(link);
+
+    return link.request.length <= 1 ? new TerminatedConcat(this, link) : new ConcatLink(this, link);
+  }
+
+  public abstract request(operation: Operation, forward?: NextLink): Observable<FetchResult> | null;
+}
+
+export function execute(link: ApolloLink, operation: GraphQLRequest): Observable<FetchResult> {
+  validateOperation(operation);
+
+  if (!operation.context) {
+    operation.context = {};
+  }
+  if (!operation.variables) {
+    operation.variables = {};
+  }
+  if (!operation.query) {
+    console.warn(`query should either be a sting or GraphQL AST`);
+    operation.query = <DocumentNode>{};
+  }
+
+  const _operation: Operation = transformOperation(operation);
+
+  return link.request(_operation) || Observable.of();
+}
+
+const toPromise = (link: ApolloLink) => {
+  return (operation: Operation, forward?: NextLink) => {
+    const observable = link.request(operation, forward);
+
+    return new Promise((resolve, reject) => {
+      observable.subscribe({
+        next: resolve,
+        error: reject,
+      });
+    });
+  };
+};
+
+export function asPromiseWrapper(link: ApolloLink | RequestHandler) {
+  if (typeof link === 'function') {
+    link = new FunctionLink(link);
+  }
+  return {
+    request: toPromise(link),
+  };
+}
+
+function transformOperation(operation) {
+  if (typeof operation.query === 'string') {
+    operation = {
+      ...operation,
+      query: parse(operation.query),
+    };
+  }
+
+  if (!operation.operationName) {
+    operation.operationName = operation.query.definitions &&
+      operation.query.definitions.length &&
+      operation.query.definitions[0].name || '';
+  }
+
+  return operation;
+}
+
+export class TerminatedLink extends ApolloLink {
```

Comment:
Also, I don't think this class is used right now, other than in a ternary expression where it doesn't provide any type safety.